### ISSUE TITLE
Experiment: GROUP BY optimization with block-local hash tables

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#21300]
If the GROUP BY key is simple (one or several fixed-size columns)
and if there is one simple aggregate function (with state up to 8 bytes),

if during aggregation the thread-local hash table exceeds the size of L2 cache, 
try to do a loop iteration with "block-local" hash table:

aggregate data into temporary hash table and them merge to the common hash table,

if temporary hash table fits L2 cache and if it's size several times less than the block size, then the method considered beneficial and can be applied in subsequent loop iterations.

Example query:

```
SELECT ClientIP AS k, count() FROM test.hits GROUP BY k ORDER BY count() DESC LIMIT 10
```

The total number of IP addresses is 
```
milovidov-desktop :) SELECT uniq(ClientIP) FROM test.hits

┌─uniq(ClientIP)─┐
│         204660 │
└────────────────┘
```

So, aggregation does not fit in L2 cache.

But while we process every block of 64K records, there are only up to 10K unique IP addresses per block:
```
SELECT
    blockNumber() AS k,
    uniq(ClientIP)
FROM test.hits
GROUP BY k

┌───k─┬─uniq(ClientIP)─┐
│ 116 │           5148 │
│  66 │           4547 │
│ 156 │           9570 │
│  46 │           7769 │
│  24 │           2270 │
│ 170 │           7363 │
│ 114 │           4825 │
│  68 │           6942 │
│ 154 │          11033 │
│  40 │           9188 │
│  30 │          10215 │
...
```

**Comment by Bulat-Ziganshin** [comment]
Can I propose an alternative approach? Radix-sort the data on the 8 bits of hash and then insert data into the main hashtable in order of these blocks. It will make accesses to the hashtable much more time-local, and in this particular case all random memory accesses while processing one block will be kept to L1$.

I think this approach has wider usage than the restrictions described in the start of your message.

**Comment by alexey-milovidov** [comment]
Do I understand the idea correctly?

For every block, make a temporary array of pairs `hash, idx` for every record
sort this array by some byte of hash,
then iterate on this array and iterate on the records indirectly by `idx`,
doing aggregation as usual.

This will make identical keys more likely to be iterated consecutively.

**Comment by Bulat-Ziganshin** [comment]
Yes, you got it right. But so far it's very rough, very generic idea.

General idea is to speed up insertion into hashtable via improving cache locality by partial presorting of the input block. And radix sort by subset of hash bits looks like the best sort algo for this usecase.

There are many variants though:
- small: 8-bit sort (preferably by bits used for table selection if 2-level hashtable is used) of blocks of about 10K records, so we try to fit all temporary data into L2$
- medium: 16-bit sort of blocks of about 1M records. It will make an extra pass over memory, but may be beneficial when hashtable became larger
- combo: 8-16 bit sort, i.e. sort data by 8 bits and then insert smaller subblocks directly, while sorting larger blocks by extra 8 bits prior to insertion
- parallel: split input blocks into subblocks based on those 8 bits, and reshuffle input data between threads (similar to the second pass of current parallel groupby algo) before insertion them into hashtables

>make a temporary array of pairs hash, idx

1. we don't need to store hash explicitly if it's easy to compute. radix sort (actually count sort) are just 2 passes over data
2. if the record is small, we can store it inline. even if it's large, we may still prefer to reorganize records during sorting. this ensures that we do only sequential memory access, which is much faster than the random one
